pkg/webhook/daemonset/validating/daemonset_create_update_handler.go

```diff
@@ -0,0 +1,163 @@
+	}
+	if spec.Selector != nil && len(spec.Selector.MatchLabels)+len(spec.Selector.MatchExpressions) == 0 {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("selector"), spec.Selector, "empty selector is invalid for daemonset"))
+	}
+
+	// Daemons typically run on more than one node, so mark Read-Write persistent disks as invalid.
+	coreVolumes, err := convertor.ConvertCoreVolumes(spec.Template.Spec.Volumes)
+	if err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Root(), spec.Template, fmt.Sprintf("Convert_v1_Volume_To_core_Volume failed: %v", err)))
+		return allErrs
+	}
+	allErrs = append(allErrs, corevalidation.ValidateReadOnlyPersistentDisks(coreVolumes, fldPath.Child("template", "spec", "volumes"))...)
+	allErrs = append(allErrs, corevalidation.ValidateNonnegativeField(int64(spec.MinReadySeconds), fldPath.Child("minReadySeconds"))...)
+
+	allErrs = append(allErrs, validateDaemonSetUpdateStrategy(&spec.UpdateStrategy, fldPath.Child("updateStrategy"))...)
+	if spec.RevisionHistoryLimit != nil {
+		// zero is a valid RevisionHistoryLimit
+		allErrs = append(allErrs, corevalidation.ValidateNonnegativeField(int64(*spec.RevisionHistoryLimit), fldPath.Child("revisionHistoryLimit"))...)
+	}
+	return allErrs
+}
+
+func validateDaemonSetUpdateStrategy(strategy *appsv1alpha1.DaemonSetUpdateStrategy, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+	switch strategy.Type {
+	case appsv1alpha1.OnDeleteDaemonSetStrategyType:
+	case appsv1alpha1.RollingUpdateDaemonSetStrategyType:
+		// Make sure RollingUpdate field isn't nil.
+		if strategy.RollingUpdate == nil {
+			allErrs = append(allErrs, field.Required(fldPath.Child("rollingUpdate"), ""))
+			return allErrs
+		}
+		if strategy.RollingUpdate.Partition != nil {
+			allErrs = append(allErrs,
+				corevalidation.ValidateNonnegativeField(
+					int64(*strategy.RollingUpdate.Partition),
+					fldPath.Child("rollingUpdate").Child("partition"))...)
+		}
+		switch strategy.RollingUpdate.Type {
+		case appsv1alpha1.StandardRollingUpdateType, appsv1alpha1.SurgingRollingUpdateType:
+		default:
+			validValues := []string{string(appsv1alpha1.StandardRollingUpdateType), string(appsv1alpha1.SurgingRollingUpdateType)}
+			allErrs = append(allErrs, field.NotSupported(fldPath.Child("rollingUpdate").Child("rollingUpdateType"), strategy, validValues))
+		}
+		allErrs = append(allErrs, validateRollingUpdateDaemonSet(strategy.RollingUpdate, fldPath.Child("rollingUpdate"))...)
+	default:
+		validValues := []string{string(appsv1alpha1.RollingUpdateDaemonSetStrategyType), string(appsv1alpha1.OnDeleteDaemonSetStrategyType)}
+		allErrs = append(allErrs, field.NotSupported(fldPath, strategy, validValues))
+	}
+	return allErrs
+}
+
+func validateRollingUpdateDaemonSet(rollingUpdate *appsv1alpha1.RollingUpdateDaemonSet, fldPath *field.Path) field.ErrorList {
```

Comment:
Can you make this method "inline" ? I don't see a strong reason to extract it out as a separate method.